### PR TITLE
Add option to set default folders with datestamps

### PR DESCRIPTION
**Pull request by TimWilms**
Allows the use of automatic folder generations based on year,month,week, day, etc.

This feature is fully optional, so this should not break existing configurations.

### CODE REVIEW

**Comment by TimWilms** [comment]
do not accept PR yet, this causes errors that i've yet to fix/discover.

- edit
had some local issues, has nothing to do with the extension. Ignore comment.

File: Classes/Hooks/DefaultUploadFolder.php [hunk 1]
```diff
@@ -131,12 +139,36 @@ protected function getDefaultUploadFolderForAllTables(
...
+            $subFolder = $userTsConfig[self::DEFAULT_UPLOAD_FOLDERS][self::DEFAULT_FOR_ALL_TABLES] ?? '';
+            $subFolder = $this->checkAndConvertForDateFormat($subFolder,$userTsConfig,self::DEFAULT_FOR_ALL_TABLES);
+        }else{
+            $subFolder = $this->checkAndConvertForDateFormat($subFolder,$defaultPageTs,self::DEFAULT_FOR_ALL_TABLES);
         }
 
         return $subFolder;
     }
+    protected function checkAndConvertForDateFormat($subFolder, $configTs, $table){
+        $table = $table ?? 'defaultForAllTables';
```
Comment (ruudsilvrants-beech) [inline Classes/Hooks/DefaultUploadFolder.php]: if(trim($subfolder) === ''){
   return $subfolder;
}
if(!isset($configArray['dateformat'] || (int)$configArray['dateformat'] !== 1){
return $subFolder;
}
$subFolder = // replace 
return $subFolder;

File: Classes/Hooks/DefaultUploadFolder.php [hunk 2]
```diff
@@ -118,10 +122,14 @@ protected function getDefaultUploadFolderForTable(
         array $defaultPageTs,
         array $userTsConfig
     ) {
-        $subFolder = $defaultPageTs['default_upload_folders.'][$table] ?? '';
-
+        $subFolder = $defaultPageTs[self::DEFAULT_UPLOAD_FOLDERS][$table] ?? '';
         if (empty($subFolder)) {
-            $subFolder = $userTsConfig['default_upload_folders.'][$table] ?? '';
+            $subFolder = $userTsConfig[self::DEFAULT_UPLOAD_FOLDERS][$table] ?? '';
+            $subFolder = $this->checkAndConvertForDateFormat($subFolder,$userTsConfig,$table);
```
Comment (ruudsilvrants-beech) [inline Classes/Hooks/DefaultUploadFolder.php]: $dateFormatConfig = $userTsConfig[self::DEFAULT_UPLOAD_FOLDERS][$table ' .'] ?? [];
$subFolder = $this->checkAndConvertForDateFormat($subFolder, $dateFormatConfig);

File: Classes/Hooks/DefaultUploadFolder.php [hunk 3]
```diff
@@ -14,9 +14,12 @@
 use TYPO3\CMS\Core\Resource\ResourceFactory;
 use TYPO3\CMS\Core\Utility\GeneralUtility;
 use TYPO3\CMS\Core\Utility\File\ExtendedFileUtility;
+use TYPO3\CMS\Extbase\Utility\DebuggerUtility;
```
Comment (ruudsilvrants-beech) [inline Classes/Hooks/DefaultUploadFolder.php]: can be removed

File: Classes/Hooks/DefaultUploadFolder.php [hunk 4]
```diff
@@ -106,9 +107,15 @@ protected function getDefaultUploadFolderForTableAndField(
...
     ) {
-        $subFolder = $defaultPageTs['default_upload_folders.'][$table.'.'][$field] ?? '';
+        $subFolder = $defaultPageTs[self::DEFAULT_UPLOAD_FOLDERS][$table . '.'][$field] ?? '';
+
+        $dateFormatConfig = $defaultPageTs[self::DEFAULT_UPLOAD_FOLDERS][$table . '.'][$field . '.'] ?? [];
+        $subFolder = $this->checkAndConvertForDateFormat($subFolder, $dateFormatConfig);
         if (empty($subFolder)) {
-            $subFolder = $userTsConfig['default_upload_folders.'][$table.'.'][$field] ?? '';
+            $subFolder = $userTsConfig[self::DEFAULT_UPLOAD_FOLDERS][$table . '.'][$field] ?? '';
+
```
Comment (ruudsilvrants-beech) [inline Classes/Hooks/DefaultUploadFolder.php]: remove empty line

File: Classes/Hooks/DefaultUploadFolder.php [hunk 5]
```diff
@@ -106,9 +107,15 @@ protected function getDefaultUploadFolderForTableAndField(
         array $defaultPageTs,
         array $userTsConfig
     ) {
-        $subFolder = $defaultPageTs['default_upload_folders.'][$table.'.'][$field] ?? '';
+        $subFolder = $defaultPageTs[self::DEFAULT_UPLOAD_FOLDERS][$table . '.'][$field] ?? '';
+
```
Comment (ruudsilvrants-beech) [inline Classes/Hooks/DefaultUploadFolder.php]: remove empty line

File: Classes/Hooks/DefaultUploadFolder.php [hunk 6]
```diff
@@ -118,25 +125,55 @@ protected function getDefaultUploadFolderForTable(
...
 
+    protected function checkAndConvertForDateFormat($subFolder, $dateFormatConfig)
+    {
+        if (trim($subFolder) === '') {
+            return $subFolder;
+        }
+        if (!isset($dateFormatConfig['dateformat']) || (int)$dateFormatConfig['dateformat'] !== 1) {
+            return $subFolder;
+        }
+        $strReplace = ['{Y}', '{y}',
```
Comment (ruudsilvrants-beech) [inline Classes/Hooks/DefaultUploadFolder.php]: The array content on a new line

File: Classes/Hooks/DefaultUploadFolder.php [hunk 7]
```diff
@@ -140,35 +142,38 @@ protected function getDefaultUploadFolderForAllTables(
...
-                $replaceWith = [
-                    date('Y'), date('y'),
-                    date('m'), date('n'),
-                    date('j'), date('d'),
-                    date('W'), date('w')
-                ];
-                $subFolder = str_replace($strplace, $replaceWith, $subFolder);
-            }
+
+    protected function checkAndConvertForDateFormat($subFolder, $dateFormatConfig)
```
Comment (FamousWolf) [inline Classes/Hooks/DefaultUploadFolder.php]: Add phpdoc annotations for the function, a return type and types for the parameters

**Review by FamousWolf — APPROVED** [review]

**Pull request merged**